components/navigation/PreventNavigation.vue

```diff
@@ -0,0 +1,124 @@
+<template>
+  <div>
+    <CModal
+      :show.sync="isOpen"
+      :close-on-backdrop="false"
+    >
+      <template #header>
+        <h5 class="modal-title">
+          <slot name="header">
+            Warning
+          </slot>
+        </h5>
+      </template>
+
+      <template>
+        <slot />
+      </template>
+
+      <template #footer>
+        <button
+          v-for="action in actions"
+          :key="action.text"
+          :ref="`actionBtn-${action.text}`"
+          type="button"
+          :class="`btn btn-${action.type || 'secondary'}`"
+          @click="actionClick(action)"
+        >
+          {{ action.text }}
+        </button>
+      </template>
+    </CModal>
+  </div>
+</template>
+<script>
+
+import { CModal } from '@coreui/vue';
+
+export default {
+  name      : 'PreventNavigation',
+  components: { CModal },
+  props     : {
+    prevent: {
+      type   : Boolean,
+      default: false,
+    },
+    actions: {
+      type    : Array,
+      required: true,
+      validator(arr) {
+        const regex = /[a-z]+/;
+        return arr.length > 0
+          && arr.every((action) => action.text && action.name
+          && regex.test(String(action.name)));
+      },
+    },
+  },
+  data() {
+    return {
+      isOpen     : false,
+      routeParams: {
+        to  : null,
+        from: null,
+        next: null,
+      },
+      unregisterMyRouterGuardListener: () => {},
+    };
+  },
+  mounted() {
+    // register system unload event (eg: if user  type a new URL)
+    window.addEventListener('beforeunload', this.preventNav);
+    // register vuejs router guard listener
+    const unregister = this.$router.beforeEach((to, from, next) => { // must be in an arrow function to save `this` closure reference
+      beforeRouteChange.call(this, { to, from, next });
+    });
```

Comment:
We used => to save this closure. pass parameters `beforeRouteChange` it  transparently wrpapped)
```suggestion
    const unregister = this.$router.beforeEach((to, from, next) => { // arrow function to save `this` closure reference
      beforeRouteChange.call(this, to, from, next);
    });
```

Reply:
DONE! changes done as suggested.